renderdoc/driver/gl/egl_hooks.cpp

```diff
@@ -647,6 +647,7 @@ void EGLHook::RegisterHooks()
   // we have to specify these with the most preferred library first. If the same function is
   // exported in multiple libraries, the function we call into will be the first one found
   LibraryHooks::RegisterLibraryHook("libGLESv3" LIBSUFFIX, NULL);
+  LibraryHooks::RegisterLibraryHook("libGLESv2_adreno" LIBSUFFIX, NULL);
```

Comment:
What's the reasoning behind needing to hook this library? I'm a bit concerned about hooking IHV libraries and things working on one system then breaking on another, since Android is so unreliable.

Since this is a public extension shouldn't it be hookable at the public libraries like anything else?

Reply:
To be honest, I am not exactly sure. For some reason I was able to hook but unable to retrieve the original function pointer to glTextureFoveationParametersQCOM until I include the adreno library directly. It is possible that the platform I am on is doing something special. I can drop this change and keep it on my fork if you feel this is not the right thing to do.

Reply:
I don't want to merge this without understanding completely why it's necessary as I definitely think it could break. I wouldn't recommend you keep it on your fork either as it might work today on your platform and break after a driver or OS update.

Do you mean that `eglGetProcAddress` returns NULL for that function? How does the application obtain the original function pointer?

Reply:
This line inside LibraryHooks::EndHookRegistration was returning NULL for glTextureFoveationParametersQCOM:

*hook.orig = dlsym(handle, hook.function.c_str());

I checked that all the other GL functions' corresponding hook.orig are NULL by that point and is relying on dlsym to retrieve the original function pointer. 

I could include the logcat, but it seems it won't let me attach a 15Mb txt file.

Reply:
Ah I remember what happened now. Inside LibraryHooks::EndHookRegistration, dlsym retrieved the hook rather than the original function pointer, therefore skipping eglGetProcAddress by the time we get to GL.PopulateWithCallback.

Reply:
I have modified LibraryHooks::EndHookRegistration and now the adreno library is no longer necessary. Please let me know if that is the right way to fix the problem!

Reply:
The guts of the hooking is even more sensitive so I definitely don't want to make any changes there without fully understanding what's going on. I can't say if it's the right way to fix the problem because it's not clear what the problem is. This change might work in this case to fix your problem but it could break other cases.

Can you give more information about exactly what went wrong and what your situation is? Are you building with interceptor-lib, and if not what happens if you do? It's technically the only supported solution because depending on the vagaries of Android platforms it might be impossible to properly support hooking without it. The only reason it's optional is because it's a significant penalty to building, so PLT hooking is provided as a 'best-effort' fallback.

When you say that `dlsym` 'retrieved the hook' what do you mean exactly by that? I don't think you mean that it's returning the hooked entry point because that's exported from renderdoc's library and this `dlsym` call isn't using `RTLD_NEXT` or `RTLD_DEFAULT` but it's using a specific library's handle. It must be returning a pointer in the library itself, and if the original pointer is NULL then it shouldn't be returning a pointer to patched hook code that will jump into renderdoc's library. What exactly was retrieved by `dlsym` and what exactly went wrong as a result?

Reply:
I am building without interceptor-lib, I can try building with it and see if that fixes the problem.

To clarify what happened: dlsym returned the address to glFramebufferFoveationConfigQCOM_renderdoc_hooked because it was using the handle to RENDERDOC_ANDROID_LIBRARY

Reply:
Yes please try with interceptor-lib and see if that makes any difference.

I see what you mean now, so the actual problem is that the `libraryHooks` vector there contains RENDERDOC_ANDROID_LIBRARY even though we only want to iterate over libraries that we're actually trying to hook.

In that case the fix would be to make sure we don't have our own library in there. E.g. by having a list of just the libraries that were registered to be hooked with `RegisterLibraryHook` and iterate only over them. In fact I'd need to check why RENDERDOC_ANDROID_LIBRARY is added to that list at all, maybe that's an artifact leftover from earlier revisions.

Reply:
In fact yes, I'd need to test it further but I'd be interested to know if your problem is also solved by commenting out the `GetHookInfo().AddLibHook(RENDERDOC_ANDROID_LIBRARY);` line entirely. Leave the `GetHookInfo().SetHooked(RENDERDOC_ANDROID_LIBRARY);` line.

FYI since interceptor-lib is a huge PITA to compile in since it requires an android build of llvm, I've put up a precompiled version of llvm [here](https://renderdoc.org/interceptor_lib_llvm.zip) you can try to use. It might save you some time.

Reply:
Running Unreal Engine 4.20, commenting out the line results in a crash inside FAndroidMisc::HasVulkanDriverSupport which has a call to dlopen, dlclose, and a few vulkan initializtion calls.

I tried turning on interceptor-lib, but I am unable to compile (on visual studio 2017) because of some unsupported flags. Is there an instruction somewhere for building Android? What do I need to do to use your precompiled version of llvm? Sorry my knowledge of cmake is pretty weak.

Reply:
I have been building android renderdoc using visual studio as my generator, it required a few minor modifications in my fork. I have tried the official compilation instructions and I can build with interceptor-lib on now! The main issue I had was trying to run cmake and make in command prompt rather than in bash shell.

I have tried your commit 90ef881 without interceptor-lib and that did indeed fixed the issue. The problem does not happen when interceptor-lib is turned on. Thank you very much for helping me with this!

Reply:
Great, I'm glad it's working now!

Building android with visual studio seems very adventurous, I personally wouldn't expect it work reliably since I don't trust android to work under normal circumstances :smile:. I'd recommend at least if you have any troubles that you're unsure about to fall back to the normal build. I use a bash shell myself rather than the command prompt, but I think it should work if you use `cmake --build` - sounds like you got it working.